Allow base-4.20

> [_15] rejecting: boring-0.2.1 (conflict: transformers => base==4.20.0.0/installed-8a80, boring => base>=4.5 && <4.20)

There is also a warning with GHC 9.10:
```
[1 of 1] Compiling Data.Boring      ( src/Data/Boring.hs, dist/build/Data/Boring.o, dist/build/Data/Boring.dyn_o )
src/Data/Boring.hs:199:12: warning: [GHC-12003] [-Wtype-equality-out-of-scope]
    • The ‘~’ operator is out of scope.
      Assuming it to stand for an equality constraint.
    • NB: ‘~’ used to be built-in syntax but now is a regular type operator
          exported from Data.Type.Equality and Prelude.
      If you are using a custom Prelude, consider re-exporting it.
    • This will become an error in a future GHC release.
    |
199 | instance a ~ b => Boring (a Eq.:~: b) where
    |            ^
```

--- a/.github/workflows/haskell-ci.yml
+++ b/.github/workflows/haskell-ci.yml
@@ -8,9 +8,9 @@
 #
 # For more information, see https://github.com/haskell-CI/haskell-ci
 #
-# version: 0.17.20231010
+# version: 0.19.20240514
 #
-# REGENDATA ("0.17.20231010",["github","cabal.project"])
+# REGENDATA ("0.19.20240514",["github","cabal.project"])
 #
 name: Haskell-CI
 on:
@@ -27,24 +27,29 @@ jobs:
     timeout-minutes:
       60
     container:
-      image: buildpack-deps:bionic
+      image: buildpack-deps:jammy
     continue-on-error: ${{ matrix.allow-failure }}
     strategy:
       matrix:
         include:
-          - compiler: ghc-9.8.1
+          - compiler: ghc-9.10.1
             compilerKind: ghc
-            compilerVersion: 9.8.1
+            compilerVersion: 9.10.1
             setup-method: ghcup
             allow-failure: false
-          - compiler: ghc-9.6.3
+          - compiler: ghc-9.8.2
             compilerKind: ghc
-            compilerVersion: 9.6.3
+            compilerVersion: 9.8.2
             setup-method: ghcup
             allow-failure: false
-          - compiler: ghc-9.4.7
+          - compiler: ghc-9.6.5
             compilerKind: ghc
-            compilerVersion: 9.4.7
+            compilerVersion: 9.6.5
+            setup-method: ghcup
+            allow-failure: false
+          - compiler: ghc-9.4.8
+            compilerKind: ghc
+            compilerVersion: 9.4.8
             setup-method: ghcup
             allow-failure: false
           - compiler: ghc-9.2.8
@@ -65,47 +70,12 @@ jobs:
           - compiler: ghc-8.8.4
             compilerKind: ghc
             compilerVersion: 8.8.4
-            setup-method: hvr-ppa
+            setup-method: ghcup
             allow-failure: false
           - compiler: ghc-8.6.5
             compilerKind: ghc
             compilerVersion: 8.6.5
-            setup-method: hvr-ppa
-            allow-failure: false
-          - compiler: ghc-8.4.4
-            compilerKind: ghc
-            compilerVersion: 8.4.4
-            setup-method: hvr-ppa
-            allow-failure: false
-          - compiler: ghc-8.2.2
-            compilerKind: ghc
-            compilerVersion: 8.2.2
-            setup-method: hvr-ppa
-            allow-failure: false
-          - compiler: ghc-8.0.2
-            compilerKind: ghc
-            compilerVersion: 8.0.2
-            setup-method: hvr-ppa
-            allow-failure: false
-          - compiler: ghc-7.10.3
-            compilerKind: ghc
-            compilerVersion: 7.10.3
-            setup-method: hvr-ppa
-            allow-failure: false
-          - compiler: ghc-7.8.4
-            compilerKind: ghc
-            compilerVersion: 7.8.4
-            setup-method: hvr-ppa
-            allow-failure: false
-          - compiler: ghc-7.6.3
-            compilerKind: ghc
-            compilerVersion: 7.6.3
-            setup-method: hvr-ppa
-            allow-failure: false
-          - compiler: ghc-7.4.2
-            compilerKind: ghc
-            compilerVersion: 7.4.2
-            setup-method: hvr-ppa
+            setup-method: ghcup
             allow-failure: false
       fail-fast: false
     steps:
@@ -113,21 +83,11 @@ jobs:
         run: |
           apt-get update
           apt-get install -y --no-install-recommends gnupg ca-certificates dirmngr curl git software-properties-common libtinfo5
-          if [ "${{ matrix.setup-method }}" = ghcup ]; then
-            mkdir -p "$HOME/.ghcup/bin"
-            curl -sL https://downloads.haskell.org/ghcup/0.1.19.5/x86_64-linux-ghcup-0.1.19.5 > "$HOME/.ghcup/bin/ghcup"
-            chmod a+x "$HOME/.ghcup/bin/ghcup"
-            "$HOME/.ghcup/bin/ghcup" install ghc "$HCVER" || (cat "$HOME"/.ghcup/logs/*.* && false)
-            "$HOME/.ghcup/bin/ghcup" install cabal 3.10.1.0 || (cat "$HOME"/.ghcup/logs/*.* && false)
-          else
-            apt-add-repository -y 'ppa:hvr/ghc'
-            apt-get update
-            apt-get install -y "$HCNAME"
-            mkdir -p "$HOME/.ghcup/bin"
-            curl -sL https://downloads.haskell.org/ghcup/0.1.19.5/x86_64-linux-ghcup-0.1.19.5 > "$HOME/.ghcup/bin/ghcup"
-            chmod a+x "$HOME/.ghcup/bin/ghcup"
-            "$HOME/.ghcup/bin/ghcup" install cabal 3.10.1.0 || (cat "$HOME"/.ghcup/logs/*.* && false)
-          fi
+          mkdir -p "$HOME/.ghcup/bin"
+          curl -sL https://downloads.haskell.org/ghcup/0.1.20.0/x86_64-linux-ghcup-0.1.20.0 > "$HOME/.ghcup/bin/ghcup"
+          chmod a+x "$HOME/.ghcup/bin/ghcup"
+          "$HOME/.ghcup/bin/ghcup" install ghc "$HCVER" || (cat "$HOME"/.ghcup/logs/*.* && false)
+          "$HOME/.ghcup/bin/ghcup" install cabal 3.10.2.0 || (cat "$HOME"/.ghcup/logs/*.* && false)
         env:
           HCKIND: ${{ matrix.compilerKind }}
           HCNAME: ${{ matrix.compiler }}
@@ -139,22 +99,13 @@ jobs:
           echo "CABAL_DIR=$HOME/.cabal" >> "$GITHUB_ENV"
           echo "CABAL_CONFIG=$HOME/.cabal/config" >> "$GITHUB_ENV"
           HCDIR=/opt/$HCKIND/$HCVER
-          if [ "${{ matrix.setup-method }}" = ghcup ]; then
-            HC=$("$HOME/.ghcup/bin/ghcup" whereis ghc "$HCVER")
-            HCPKG=$(echo "$HC" | sed 's#ghc$#ghc-pkg#')
-            HADDOCK=$(echo "$HC" | sed 's#ghc$#haddock#')
-            echo "HC=$HC" >> "$GITHUB_ENV"
-            echo "HCPKG=$HCPKG" >> "$GITHUB_ENV"
-            echo "HADDOCK=$HADDOCK" >> "$GITHUB_ENV"
-            echo "CABAL=$HOME/.ghcup/bin/cabal-3.10.1.0 -vnormal+nowrap" >> "$GITHUB_ENV"
-          else
-            HC=$HCDIR/bin/$HCKIND
-            echo "HC=$HC" >> "$GITHUB_ENV"
-            echo "HCPKG=$HCDIR/bin/$HCKIND-pkg" >> "$GITHUB_ENV"
-            echo "HADDOCK=$HCDIR/bin/haddock" >> "$GITHUB_ENV"
-            echo "CABAL=$HOME/.ghcup/bin/cabal-3.10.1.0 -vnormal+nowrap" >> "$GITHUB_ENV"
-          fi
-
+          HC=$("$HOME/.ghcup/bin/ghcup" whereis ghc "$HCVER")
+          HCPKG=$(echo "$HC" | sed 's#ghc$#ghc-pkg#')
+          HADDOCK=$(echo "$HC" | sed 's#ghc$#haddock#')
+          echo "HC=$HC" >> "$GITHUB_ENV"
+          echo "HCPKG=$HCPKG" >> "$GITHUB_ENV"
+          echo "HADDOCK=$HADDOCK" >> "$GITHUB_ENV"
+          echo "CABAL=$HOME/.ghcup/bin/cabal-3.10.2.0 -vnormal+nowrap" >> "$GITHUB_ENV"
           HCNUMVER=$(${HC} --numeric-version|perl -ne '/^(\d+)\.(\d+)\.(\d+)(\.(\d+))?$/; print(10000 * $1 + 100 * $2 + ($3 == 0 ? $5 != 1 : $3))')
           echo "HCNUMVER=$HCNUMVER" >> "$GITHUB_ENV"
           echo "ARG_TESTS=--enable-tests" >> "$GITHUB_ENV"
@@ -213,14 +164,14 @@ jobs:
       - name: install cabal-docspec
         run: |
           mkdir -p $HOME/.cabal/bin
-          curl -sL https://github.com/phadej/cabal-extras/releases/download/cabal-docspec-0.0.0.20230517/cabal-docspec-0.0.0.20230517-x86_64-linux.xz > cabal-docspec.xz
-          echo '3b31bbe463ad4d671abbc103db49628562ec48a6604cab278207b5b6acd21ed7  cabal-docspec.xz' | sha256sum -c -
+          curl -sL https://github.com/phadej/cabal-extras/releases/download/cabal-docspec-0.0.0.20240414/cabal-docspec-0.0.0.20240414-x86_64-linux.xz > cabal-docspec.xz
+          echo '2d18a3f79619e8ec5f11870f926f6dc2616e02a6c889315b7f82044b95a1adb9  cabal-docspec.xz' | sha256sum -c -
           xz -d < cabal-docspec.xz > $HOME/.cabal/bin/cabal-docspec
           rm -f cabal-docspec.xz
           chmod a+x $HOME/.cabal/bin/cabal-docspec
           cabal-docspec --version
       - name: checkout
-        uses: actions/checkout@v3
+        uses: actions/checkout@v4
         with:
           path: source
       - name: initial cabal.project for sdist
@@ -244,11 +195,11 @@ jobs:
           touch cabal.project
           touch cabal.project.local
           echo "packages: ${PKGDIR_boring}" >> cabal.project
-          if [ $((HCNUMVER >= 80200)) -ne 0 ] ; then echo "package boring" >> cabal.project ; fi
-          if [ $((HCNUMVER >= 80200)) -ne 0 ] ; then echo "    ghc-options: -Werror=missing-methods" >> cabal.project ; fi
+          echo "package boring" >> cabal.project
+          echo "    ghc-options: -Werror=missing-methods" >> cabal.project
           cat >> cabal.project <<EOF
           EOF
-          $HCPKG list --simple-output --names-only | perl -ne 'for (split /\s+/) { print "constraints: $_ installed\n" unless /^(boring)$/; }' >> cabal.project.local
+          $HCPKG list --simple-output --names-only | perl -ne 'for (split /\s+/) { print "constraints: any.$_ installed\n" unless /^(boring)$/; }' >> cabal.project.local
           cat cabal.project
           cat cabal.project.local
       - name: dump install plan
@@ -256,7 +207,7 @@ jobs:
           $CABAL v2-build $ARG_COMPILER $ARG_TESTS $ARG_BENCH --dry-run all
           cabal-plan
       - name: restore cache
-        uses: actions/cache/restore@v3
+        uses: actions/cache/restore@v4
         with:
           key: ${{ runner.os }}-${{ matrix.compiler }}-${{ github.sha }}
           path: ~/.cabal/store
@@ -296,7 +247,7 @@ jobs:
           $CABAL v2-build $ARG_COMPILER --disable-tests --disable-benchmarks --constraint='boring -tagged' --dependencies-only -j2 all
           $CABAL v2-build $ARG_COMPILER --disable-tests --disable-benchmarks --constraint='boring -tagged' all
       - name: save cache
-        uses: actions/cache/save@v3
+        uses: actions/cache/save@v4
         if: always()
         with:
           key: ${{ runner.os }}-${{ matrix.compiler }}-${{ github.sha }}

--- a/boring-instances/boring-instances.cabal
+++ b/boring-instances/boring-instances.cabal
@@ -1,3 +1,4 @@
+cabal-version:      2.2
 name:               boring-instances
 version:            0.2
 synopsis:           Boring and Absurd types: instances
@@ -20,7 +21,6 @@ maintainer:         Oleg.Grenrus <oleg.grenrus@iki.fi>
 copyright:          (c) 2017-2021 Oleg Grenrus
 category:           Data
 build-type:         Simple
-cabal-version:      >=1.10
 tested-with:
   GHC ==7.8.4
    || ==7.10.3

--- a/boring/ChangeLog.md
+++ b/boring/ChangeLog.md
@@ -1,5 +1,9 @@
 # Revision history for boring
 
+## 0.2.2
+
+- Support GHC-8.6.5...9.10.1
+
 ## 0.2.1
 
 -  Add instances for `SNat`, `SSymbol`, and `SChar`

--- a/boring/boring.cabal
+++ b/boring/boring.cabal
@@ -1,17 +1,17 @@
+cabal-version:      2.2
 name:               boring
-version:            0.2.1
-x-revision:         1
+version:            0.2.2
 synopsis:           Boring and Absurd types
 description:
   * @Boring@ types are isomorphic to @()@.
-  . 
+  .
   * @Absurd@ types are isomorphic to @Void@.
   .
   See [What does () mean in Haskell -answer by Conor McBride](https://stackoverflow.com/questions/33112439/what-does-mean-in-haskell/33115522#33115522)
 
 homepage:           https://github.com/phadej/boring
 bug-reports:        https://github.com/phadej/boring/issues
-license:            BSD3
+license:            BSD-3-Clause
 license-file:       LICENSE
 author:             Oleg Grenrus <oleg.grenrus@iki.fi>
 maintainer:         Oleg.Grenrus <oleg.grenrus@iki.fi>
@@ -19,23 +19,16 @@ copyright:          (c) 2017-2021 Oleg Grenrus
 category:           Data
 build-type:         Simple
 extra-source-files: ChangeLog.md
-cabal-version:      >=1.10
 tested-with:
-  GHC ==7.4.2
-   || ==7.6.3
-   || ==7.8.4
-   || ==7.10.3
-   || ==8.0.2
-   || ==8.2.2
-   || ==8.4.4
-   || ==8.6.5
+  GHC ==8.6.5
    || ==8.8.4
    || ==8.10.7
    || ==9.0.2
    || ==9.2.8
-   || ==9.4.7
-   || ==9.6.3
-   || ==9.8.1
+   || ==9.4.8
+   || ==9.6.5
+   || ==9.8.2
+   || ==9.10.1
 
 source-repository head
   type:     git
@@ -54,24 +47,13 @@ flag tagged
 library
   exposed-modules:  Data.Boring
   build-depends:
-      base          >=4.5 && <4.20
-    , transformers  >=0.3 && <0.7
+      base          >=4.12.0.0 && <4.21
+    , transformers  >=0.5.6.2  && <0.7
 
   if impl(ghc <7.6)
     build-depends: ghc-prim
-  
-  if !impl(ghc >=7.8)
-    build-depends: type-equality >=1 && <1.1
 
-  if !impl(ghc >=7.10)
-    build-depends: void >=0.7.3 && <0.8
-
-  if !impl(ghc >=8.0)
-    build-depends:
-        semigroups           >=0.18.5 && <0.21
-      , transformers-compat  >=0.5    && <0.8
-
-  if flag(tagged) || !impl(ghc >=7.8)
+  if flag(tagged)
     build-depends: tagged >=0.8.6 && <0.9
 
   other-extensions:

--- a/boring/src/Data/Boring.hs
+++ b/boring/src/Data/Boring.hs
@@ -4,9 +4,7 @@
 {-# LANGUAGE GADTs             #-}
 {-# LANGUAGE Trustworthy       #-}
 {-# LANGUAGE TypeOperators     #-}
-#if __GLASGOW_HASKELL >= 708
 {-# LANGUAGE EmptyCase         #-}
-#endif
 -- | 'Boring' and 'Absurd' classes. One approach.
 --
 -- Different approach would be to have
@@ -79,18 +77,12 @@ import GHC.Generics
 
 import qualified Data.Void as V
 
-#if __GLASGOW_HASKELL >= 708
 import qualified Data.Coerce        as Co
 import qualified Data.Type.Coercion as Co
-#else
-import Prelude (seq, error)
-#endif
 
 import qualified Data.Type.Equality as Eq
 
-#if MIN_VERSION_base(4,10,0)
 import qualified Type.Reflection as Typeable
-#endif
 
 #if MIN_VERSION_base(4,18,0)
 import qualified GHC.TypeLits as TypeLits
@@ -189,26 +181,20 @@ instance Absurd a => Boring [a] where
 instance Absurd a => Boring (Maybe a) where
     boring = Nothing
 
-#if __GLASGOW_HASKELL >= 708
 -- | Coercibility is 'Boring' too.
 instance Co.Coercible a b => Boring (Co.Coercion a b) where
     boring = Co.Coercion
-#endif
 
 -- | Homogeneous type equality is 'Boring' too.
 instance a ~ b => Boring (a Eq.:~: b) where
     boring = Eq.Refl
 
-# if MIN_VERSION_base(4,10,0)
 -- | Heterogeneous type equality is 'Boring' too.
 instance a Eq.~~ b => Boring (a Eq.:~~: b) where
     boring = Eq.HRefl
-# endif
 
-#if MIN_VERSION_base(4,10,0)
 instance Typeable.Typeable a => Boring (Typeable.TypeRep a) where
     boring = Typeable.typeRep
-#endif
 
 #if MIN_VERSION_base(4,18,0)
 instance TypeLits.KnownChar n => Boring (TypeLits.SChar n) where
@@ -295,11 +281,7 @@ instance Absurd a => Absurd (Tagged b a) where
 -------------------------------------------------------------------------------
 
 instance Absurd (V1 p) where
-#if __GLASGOW_HASKELL >= 708
     absurd v = case v of {}
-#else
-    absurd v = v `seq` error "absurd @(V1 p)"
-#endif
 
 instance Absurd c => Absurd (K1 i c p) where
     absurd = absurd . unK1
@@ -399,11 +381,7 @@ class GAbsurd f where
     gabsurd :: f a -> b
 
 instance GAbsurd V1 where
-#if __GLASGOW_HASKELL >= 708
     gabsurd x = case x of {}
-#else
-    gabsurd x = x `seq` error "gabsurd @V1"
-#endif
 
 instance GAbsurd f => GAbsurd (M1 i c f) where
     gabsurd (M1 x) = gabsurd x

